File: web/schnorr.ts
import { CalculateCofactor, GetCurve } from "./ecc_ops";
import { Hash } from "./hash";
import { CompareTo, GenerateKey, ModuloN } from "./ops";
import { SchnorrZKP, SupportedCurves } from "./types";
import { ProjPointType } from "@noble/curves/abstract/weierstrass";

async function GenerateZKPGProvided(
    curve: SupportedCurves, 
    g: ProjPointType<bigint>, 
    n: bigint, 
    x: bigint, 
    X: ProjPointType<bigint>, 
    prover: string
): Promise<SchnorrZKP> {
    const v = await GenerateKey(curve)
    const V = g.multiply(v);
    const h = await Hash(g.toRawBytes(), V.toRawBytes(), X.toRawBytes(), prover);
    let r = x * h;
    r = v - r;
    r = ModuloN(r, n);
    return { V, r };
};

async function VerifyZKP(
    curve: SupportedCurves,
    generator: ProjPointType<bigint>,
    X: ProjPointType<bigint>,
    zkp: SchnorrZKP,
    prover: string
): Promise<boolean> {

    const h = await Hash(generator.toRawBytes(), zkp.V.toRawBytes(), X.toRawBytes(), prover);
    if (!X || !zkp.V || !zkp.r) return false;

    try { X.assertValidity(); } 
    catch { return false; }

    const { x: xX, y: yY } = X.toAffine();
    if (xX === null || yY === null) return false;

    const curveParams = GetCurve(curve);
    if (CompareTo(xX, BigInt(0)) === -1 || CompareTo(xX, curveParams.CURVE.n - BigInt(1)) === 1) return false;
    if (CompareTo(yY, BigInt(0)) === -1 || CompareTo(yY, curveParams.CURVE.n - BigInt(1)) === 1) return false;

    const xXh = X.multiply(CalculateCofactor(curve));
    const { x: xXhX, y: xXhY } = xXh.toAffine();
    if (xXhX === null || xXhY === null) return false;
    if (xXhX === BigInt(0) && xXhY === BigInt(0)) return false;

    const gRxhmn = generator.multiply(zkp.r).add(X.multiply(ModuloN(h, curveParams.CURVE.n)));
    return zkp.V.equals(gRxhmn);
}

export {
    GenerateZKPGProvided,
    VerifyZKP
}

File: pkg/crypto/schnorr.go
package crypto

import (
	"crypto/elliptic"
	"math/big"
)

type SchnorrZKP struct {
	V []byte
	R *big.Int
}

func GenerateZKP(
	generator elliptic.Curve,
	n *big.Int,
	x *big.Int,
	X []byte,
	userID string,
) *SchnorrZKP {
	g := GetG(generator)
	return GenerateZKPGProvided(generator, g, n, x, X, userID)
}

func GenerateZKPGProvided(
	curve elliptic.Curve,
	g []byte, // No pointer needed
	n *big.Int,
	x *big.Int,
	X []byte,
	prover string,
) *SchnorrZKP {
	v := GenerateKey(curve)
	V, err := MultiplyPoint(curve, &g, v)
	if err != nil {
		panic(err)
	}
	h := Hash(g, V, X, prover)
	r := Multiply(x, h)
	r = new(big.Int).Sub(v, r)
	r = ModuloN(r, n)
	return &SchnorrZKP{V, r}
}

func VerifyZKP(
	curve elliptic.Curve,
	generator []byte,
	X []byte,
	zkp SchnorrZKP,
	prover string,
) bool {
	h := Hash(generator, zkp.V, X, prover)

	if X == nil || zkp.V == nil || zkp.R == nil {
		return false
	}

	xX, xY := elliptic.UnmarshalCompressed(curve, X)
	if IsInfinity(xX, xY) {
		return false
	}

	if xX.Cmp(big.NewInt(0)) == -1 || xX.Cmp(new(big.Int).Sub(curve.Params().N, big.NewInt(1))) == 1 {
		return false
	}

	if xY.Cmp(big.NewInt(0)) == -1 || xY.Cmp(new(big.Int).Sub(curve.Params().N, big.NewInt(1))) == 1 {
		return false
	}

	if !curve.IsOnCurve(xX, xY) {
		return false
	}

	xXh, err := MultiplyPoint(curve, &X, CalculateCofactor(curve))
	if err != nil {
		return false
	}
	xXhX, xXhY := elliptic.UnmarshalCompressed(curve, xXh)
	if IsInfinity(xXhX, xXhY) {
		return false
	}

	gr, err := MultiplyPoint(curve, &generator, zkp.R)
	if err != nil {
		return false
	}

	Xhn, err := MultiplyPoint(curve, &X, ModuloN(h, curve.Params().N))
	if err != nil {
		return false
	}

	gRxhmn, err := AddPoints(curve, gr, Xhn)
	if err != nil {
		return false
	}

	return PointsEqual(curve, zkp.V, gRxhmn)
}
